test/jdk/com/sun/net/httpserver/HttpsParametersClientAuthTest.java

```diff
@@ -0,0 +1,513 @@
+
+        private static final int NO_RESPONSE_BODY = -1;
+
+        @Override
+        public void handle(final HttpExchange exchange) throws IOException {
+            System.out.println("responding to request: " + exchange.getRequestURI());
+            exchange.sendResponseHeaders(200, NO_RESPONSE_BODY);
+        }
+    }
+
+    private record SSLCtx(SSLContext sslContext, DelegatingKeyManager keyManager) {
+    }
+
+    private static final class DelegatingKeyManager extends X509ExtendedKeyManager {
+
+        private final X509ExtendedKeyManager underlying;
+        private boolean clientAuthInitiated;
+
+        // km can be null
+        private DelegatingKeyManager(final X509ExtendedKeyManager km) {
+            this.underlying = km;
+        }
+
+        @Override
+        public String chooseEngineClientAlias(String[] keyType, Principal[] issuers,
+                                              SSLEngine engine) {
+            clientAuthInitiated = true; // keep track that client alias selection was requested
+            if (this.underlying != null) {
+                return this.underlying.chooseEngineClientAlias(keyType, issuers, engine);
+            }
+            return null;
+        }
+
+        @Override
+        public String chooseEngineServerAlias(String keyType, Principal[] issuers,
+                                              SSLEngine engine) {
+            if (this.underlying != null) {
+                return this.underlying.chooseEngineServerAlias(keyType, issuers, engine);
+            }
+            return null;
+        }
+
+        @Override
+        public String[] getClientAliases(String keyType, Principal[] issuers) {
+            if (this.underlying != null) {
+                return this.underlying.getClientAliases(keyType, issuers);
+            }
+            return null;
+        }
+
+        @Override
+        public String chooseClientAlias(String[] keyType, Principal[] issuers, Socket socket) {
+            clientAuthInitiated = true; // keep track that client alias selection was requested
+            if (this.underlying != null) {
+                return this.underlying.chooseClientAlias(keyType, issuers, socket);
+            }
+            return null;
+        }
+
+        @Override
+        public String[] getServerAliases(String keyType, Principal[] issuers) {
+            if (this.underlying != null) {
+                return this.underlying.getServerAliases(keyType, issuers);
+            }
+            return null;
+        }
+
+        @Override
+        public String chooseServerAlias(String keyType, Principal[] issuers, Socket socket) {
+            if (this.underlying != null) {
+                return this.underlying.chooseServerAlias(keyType, issuers, socket);
+            }
+            return null;
+        }
+
+        @Override
+        public X509Certificate[] getCertificateChain(String alias) {
+            clientAuthInitiated = true; // keep track that client certs was requested
```

Comment:
nit: this is not necessary - `getCertificateChain` is only called if `choose*Alias` returns a non-null value, and both the `chooseClientAlias` methods set `clientAuthInitiated = true` already.